### PR TITLE
feat(lightning): add weights_only parameter support for Lightning 2.6.0

### PR DESCRIPTION
**Pull request by jet-tong**
## Description

**Problem**: Lightning 2.6.0 [added `weights_only` parameter](https://github.com/Lightning-AI/pytorch-lightning/pull/21072) to `CheckpointIO.load_checkpoint()`, causing TypeError when using `S3LightningCheckpoint`.

**Solution**: Add `weights_only: Optional[bool] = None` parameter and pass it to `torch.load()`, and add Lightning version-aware default for `weights_only` to avoid breaking changes.

This PR adds support for the `weights_only` parameter introduced in PyTorch Lightning 2.6.0:
- Support `weights_only` parameter
  - Add `weights_only` parameter to `S3LightningCheckpoint.load_checkpoint()
  - Add version aware default (False for Lightning <2.6, None for Lightning >=2.6)
  - Pass weights_only parameter to `torch.load()` instead of hardcoded False
- New `weights_only` tests (conditional on Lightning 2.6.0+)
  - Add new unit tests to cover None/True/False `weights_only` values, and version aware default.
  - Add integration test to test for `trainer.fit()` -> `S3LightningCheckpoint` path

Fixes compatibility with PyTorch Lightning 2.6.0, resolving https://github.com/awslabs/s3-connector-for-pytorch/issues/387. Following their [`weights_only` changes](https://github.com/Lightning-AI/pytorch-lightning/pull/21072), you can now explicitly pass the parameter through `Trainer.{fit,validate,test,predict}`.

No breaking changes - all potential breaking changes should be mitigated in this PR. If necessary, users may set PyTorch environment variable by `export TORCH_FORCE_WEIGHTS_ONLY_LOAD=0` for `weights_only` to behave as `False`.

Note:
- PyTorch <2.6 `weights_only=None` used to behave as False before (hard coded), but will now behave as True. See [PyTorch Serialization Semantics docs](https://docs.pytorch.org/docs/2.9/notes/serialization.html#torch-load-with-weights-only-true) for more details. 
- Note Lightning <2.6 also defaulted to `weights_only=False` before [[see cloud_io.py in lightning 2.5](https://github.com/Lightning-AI/pytorch-lightning/blob/5b3bb743100ace12e8326fb5ad5a4c5eb5b82887/src/lightning/fabric/utilities/cloud_io.py#L37)].

### Context: `weights_only` parameter

`weights_only` parameter controls PyTorch's unpickling security ([ref: PyTorch Serialization Semantics docs](https://docs.pytorch.org/docs/2.9/notes/serialization.html#torch-load-with-weights-only-true)):
- When `True`, restricts `torch.load()` to only load tensors, primitive types, and safe classes - preventing arbitrary code execution from untrusted checkpoints. 
- When `False`, allows full pickle functionality but poses security risks with untrusted files.
- When `None`, defers to PyTorch's default behavior (which changed from False to True in PyTorch 2.6). 

We no longer default `weights_only=False` since Lightning 2.6.0 now lets PyTorch decide on default behavior.

Referencing outdated code comments (in PR #306):
```
We PREVIOUSLY defaulted to weights_only=False to:
1. Maintain backwards compatibility with older PyTorch versions where this was the default behavior
2. Match PyTorch Lightning's implementation strategy for consistent behavior
Reference: https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/fabric/utilities/cloud_io.py#L36
```

### Backwards Compatibility

Example flow: 
`Trainer.fit(..., weights_only: Optional[bool]=None)`
-> `S3LightningCheckpoint.load_checkpoint(..., weights_only: Optional[bool] = None)`
-> `torch.load(..., weights_only: Optional[bool] = None)`

Problems:
1. Lightning 2.6 introduces `weights_only` parameter in `Trainer.{fit,validate,test,predict}` used in `CheckpointIO.load_checkpoint` to control `torch.load` behavior. 
**Mitigation**: add optional bool parameter in `S3LightningCheckpoint.load_checkpoint`, and use version-aware `weights_only` default as in next point.
2. PyTorch <2.6 `weights_only` default was False, >=2.6 uses True, but Lightning <2.6 overrides defaults to `weights_only=False` [[reference here](https://github.com/Lightning-AI/pytorch-lightning/blob/5b3bb743100ace12e8326fb5ad5a4c5eb5b82887/src/lightning/fabric/utilities/cloud_io.py#L37)]. There would be behavioral changes for specific case: PyTorch >2.6 & Lightning <2.6 & this PR -> `weights_only` would behave as `True` instead of `False`. 
**Mitigation**: Use version-aware `weights_only` default: False for Lightning <2.6, None for Lightning >=2.6. This avoids breaking change.
3. PyTorch <2.4 uses non-optional bool weights_only.
**Mitigation**: Our new default `None` acts as `False` in `if weights_only:` checks (default for PyTorch <2.6 or Lightning <2.6) for backwards compatibility.

Also checked:
- Python 3.8 uses lightning 2.3.3 and torch 2.4.1 and tests pass. 
- Previous unit tests and integration test also passes with PyTorch 2.2.0 and Lightning 2.1.0. 
- `torch.load(..., weights_only=None)` works correctly for PyTorch 2.0 (where PyTorch < 2.4 uses non-optional bool)

- [x] I have updated the CHANGELOG or README if appropriate

## Related items

- Lightning `weights_only` changes PR: https://github.com/Lightning-AI/pytorch-lightning/pull/21072
- Self-cut Issue (to close during version release): https://github.com/awslabs/s3-connector-for-pytorch/issues/387
- This PR is the follow-up for mitigation PR: https://github.com/awslabs/s3-connector-for-pytorch/pull/389
- Previous PR which set S3LightningCheckpoint `weights_only` to False: https://github.com/awslabs/s3-connector-for-pytorch/pull/306

## Testing

Unit and integration tests pass. See Backwards Compatibility section for additional checks.

--------
By submitting this pull request, I confirm that my contribution is made under the terms of BSD 3-Clause License and I agree to the terms of the [LICENSE](https://github.com/awslabs/s3-connector-for-pytorch/blob/main/LICENSE).

### CODE REVIEW

File: s3torchconnector/src/s3torchconnector/lightning/s3_lightning_checkpoint.py [hunk 1]
```diff
@@ -74,11 +77,10 @@ def load_checkpoint(
...
 
-        # Explicitly set weights_only=False to:
-        # 1. Maintain backwards compatibility with older PyTorch versions where this was the default behavior
-        # 2. Match PyTorch Lightning's implementation strategy for consistent behavior
-        # Reference: https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/fabric/utilities/cloud_io.py#L36
-        return torch.load(s3reader, map_location, weights_only=False)  # type: ignore
+        # We no longer default weights_only=False since Lightning 2.6.0 lets PyTorch decide on default behavior
+        # weights_only can be passed to load_checkpoint through Trainer.{fit,validate,test,predict}
+
+        return torch.load(s3reader, map_location, weights_only=weights_only)  # type: ignore
```
[the remark this answers is not in the document]
Comment (jet-tong) [reply s3torchconnector/src/s3torchconnector/lightning/s3_lightning_checkpoint.py]: Fortunately None acts as False, which is the default in those older PyTorch versions. We can simply add comment in code for this.

**Review by muddyfish — APPROVED** [review]

File: s3torchconnector/src/s3torchconnector/lightning/s3_lightning_checkpoint.py [hunk 2]
```diff
@@ -55,12 +56,16 @@ def load_checkpoint(
...
         map_location: Optional[Any] = None,
+        weights_only: Optional[bool] = None,
     ) -> Dict[str, Any]:
         """Load checkpoint from an S3 location when resuming or loading ckpt for test/validate/predict stages.
 
         Args:
             path (str): S3 uri to checkpoint
             map_location: A function, :class:`torch.device`, string or a dict specifying how to remap storage locations.
+            weights_only: If True, only loads tensors and primitive types (safer). If False, allows loading
+                arbitrary Python objects (less secure). If None, uses PyTorch default behavior.
```
Comment (muddyfish) [inline s3torchconnector/src/s3torchconnector/lightning/s3_lightning_checkpoint.py]: This comment may want updating - use pytorch lightning default behaviour
